src/main/java/duke/command/Command.java

```diff
@@ -0,0 +1,38 @@
+package duke.command;
+
+import duke.exception.DukeExecutionException;
+import duke.exception.DukeInvalidCommandException;
+import duke.storage.Storage;
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+/**
+ * The Command base class.
+ */
+public abstract class Command {
+    /**
+     * Checks whether this Command is valid.
+     * @param tasks the current tasks
+     * @throws DukeInvalidCommandException if this Command is invalid
+     */
+    protected abstract void check(final TaskList tasks) throws DukeInvalidCommandException;
```

Comment:
it's not immediately clear why checking if the command is valid -- involving everything from command syntax to missing arguments to invalid operations depending on the state of the task list -- should be encapsulated in the same method and take the TaskList as an argument. however, you seem to have implemented it quite well! I like how you've made all these various exceptions DukeInvalidCommandException. 👍